### ISSUE TITLE
Type changes not detected in composite mode

### ISSUE THREAD

**Issue by drrnkin** [microsoft/TypeScript#51648]
# Bug Report

### 🔎 Search Terms
composite type check tsconfig.tsbuildinfo

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about composite mode type checking.
- Discovered in 4.7.4.  Still an issue in current latest (4.9.3).

### 💻 Code

Can't use Playground or TypeScript Workbench since it appears to be a cache issue with tsconfig.tsbuildinfo.  Code must be built twice to see the issue.

Please clone https://github.com/drrnkin/typescript-composite-type-checking.  It shows that the compiler fails to detect a compilation error.  This only happens in composite mode when a third-party library has a breaking change.

- In both the main and child folders, run "yarn" to install each.
- In the main folder, run "yarn build" to build both projects in composite mode.
- In the child folder, run "yarn oldjquery" to change type info.
- In the main folder, run "yarn build" again and note that it doesn't detect the breaking change.

It's not about jQuery; that's just an example.  It's about any 3rd party package where type information changes.

In the example, Tween exists in jQuery 3, but not 2.  When compiling with 3 first it builds correctly.  When switching to 2, the build step reports no errors but misses the fact that there is a compilation error in the child project.  Note that compiling the child project directly (with "yarn tsc") detects the error, but it is missed in composite mode.

```ts
export function typeCheckingExample() {
    console.log("jQuery.Tween", jQuery.Tween); // Tween doesn't exist in jQuery 2.x, but exists in 3.x
}
```

### 🙁 Actual behavior

Code that should not compile is not detected and reported.

### 🙂 Expected behavior

Should report an error.

**Comment by sheetalkamat** [comment]
This issue is working as intended. `tsc -b` only looks at all the input and output file timestamps that are defined by your config file. It does not look at dependencies and other things. This is the output from running `tsc -b -v`


```
$ tsc -b -v
[2:17:53 PM] Projects in this build: 
    * ../child/tsconfig.json
    * tsconfig.json

[2:17:53 PM] Project '../child/tsconfig.json' is up to date because newest input '../child/child.ts' is older than output '../child/tsconfig.tsbuildinfo'

[2:17:53 PM] Project 'tsconfig.json' is up to date because newest input 'main.ts' is older than output 'tsconfig.tsbuildinfo'
```

**Comment by dimikot** [comment]
@sheetalkamat Can tsc extract the maximum timestamp from at least the files mentioned in the existing buildinfo file, in fileNames section? Getting timestamps shouldn't slow the process down too much. I saw a [work-around](https://github.com/microsoft/TypeScript/issues/53684#issuecomment-1499430841) recommended by @RyanCavanaugh to use `--force`, but in our case, using `--force` increases the no-changes build time from ~26 seconds to ~49 seconds - i.e. it almost doubles it! Possibly `--force` does some more stuff rather than disabling the timestamp cutoff check (I don't believe that getting timestamps of several thousand files can take 25+ seconds on my laptop; in fact we have some similar logic in one of other build steps, and it enumerates all the files and their timestamps, including node_modules, in just several seconds).

Context: in my case, all the files in src are older than the mtime of dist/tsconfig.tsbuildinfo, so tsc says: `Project 'tsconfig.json' is up to date because newest input 'src/commands/github/GitHubBranchNewIO.ts' is older than output 'dist/sconfig.tsbuildinfo'`. At the same file, some of other files imported (in node_modules) are newer than tsbuildinfo file, but tsc doesn't recognize them, despite of them being mentioned in dist/tsconfig.tsbuildinfo file (since they're not under `"include": ["src/**/*"]` I think).

What I did for the diagnostics: I extracted `fileNames` from dist/tsconfig.tsbuildinfo to /tmp/files (220 filenames) and then sorted them by mtime using:

```
(cd dist && while read filename; do stat -c "%y %n" "$filename"; done < /tmp/files | sort)
```

Then I ran:

```
../../../node_modules/.bin/tsc --build --verbose
```

![CleanShot 2023-11-30 at 18 31 19@2x](https://github.com/microsoft/TypeScript/assets/65643/3705c554-c9ac-4a10-ad63-f1a1a9e97e86)

We can clearly see that GitHubBranchNewIO.ts is not the newest file amongst fileNames mentioned in the buildinfo file (it's just the newest one in src).
